plugin/client_rpc.go

```diff
@@ -395,26 +395,31 @@ func (g *hooksRPCClient) FileWillBeUploaded(c *Context, info *model.FileInfo, fi
 		serveIOReader(file, uploadedFileConnection)
 	}()
 
+	replacementDone := make(chan bool)
 	replacementFileStreamId := g.muxBroker.NextId()
 	go func() {
+		defer close(replacementDone)
+
 		replacementFileConnection, err := g.muxBroker.Accept(replacementFileStreamId)
 		if err != nil {
 			g.log.Error("Plugin failed to serve replacement file stream. MuxBroker could not Accept connection", mlog.Err(err))
 			return
 		}
 		defer replacementFileConnection.Close()
-		if _, err := io.Copy(output, replacementFileConnection); err != nil && err != io.EOF {
+		if _, err := io.Copy(output, replacementFileConnection); err != nil {
 			g.log.Error("Error reading replacement file.", mlog.Err(err))
 		}
 	}()
 
 	_args := &Z_FileWillBeUploadedArgs{c, info, uploadedFileStreamId, replacementFileStreamId}
 	_returns := &Z_FileWillBeUploadedReturns{A: _args.B}
-	if g.implemented[FileWillBeUploadedId] {
-		if err := g.client.Call("Plugin.FileWillBeUploaded", _args, _returns); err != nil {
-			g.log.Error("RPC call FileWillBeUploaded to plugin failed.", mlog.Err(err))
-		}
+	if err := g.client.Call("Plugin.FileWillBeUploaded", _args, _returns); err != nil {
+		g.log.Error("RPC call FileWillBeUploaded to plugin failed.", mlog.Err(err))
 	}
+
+	// Ensure the io.Copy from the replacementFileConnection above completes.
+	<-replacementDone
```

Comment:
So the interesting question here is: can we deadlock? The server implementation of `FileWillBeUploaded` below guarantees we `Close` the `replacementFileConnection` only if we successfully `Dial` it, and the `go-plugin` implementation of `Accept` in the goroutine above times out after 5 seconds if the connection is not accepted, so this shouldn't be a concern.

Reply:
Not an expert of [go-plugin](https://github.com/hashicorp/go-plugin) but from what I can see I think it should be safe.

Personal curiosity: any particular reason why you chose to use a `channel` as opposed to `sync.WaitGroup` to achieve synchronization?

Reply:
I could go either way in this case, but thought that it might be weird to use a `WaitGroup` on only one of the spawned goroutines, when usually it's used to coordinate more than one. Of course, I'm still waiting for the first goroutine implicitly, but figured a channel named explicitly would suffice. 0/5.

Reply:
Sounds good. I appreciate the explicitness with the channel name :+1: